---
title: "Class 11: Structural Bioinformatic pt2"
author: "Joseph Girgiss (PID: A17388247)"
format: pdf
toc: TRUE
---

## AlphaFold Database

The EBI maintains the largest database of AlphaFold structure prediction models at: https://alphafold.ebi.ac.uk/

From last class (before Halloween) we saw that the PDB had 244,290 (Oct 2025). 

The total number of protein sequences in UniProtKB is 199,579,901. 

> **Key Point**: This is a tiny fraction of sequence space that has structural coverage (0.12%). 

```{r}
244290/199579901 * 100
```

AFDB is attempting to address this gap... 

There are two "Quality Scores" from AlphaFold. One for residues (i.e. each amino acid) called **plDDT** score. The second Predicted Alignment Error (**PAE**) which measures the confidence in the relative position of two residues (i.e. a score for every pair of residues). 

## Generating your own structure predictions

![Figure of 1 and 5 generated HIV-PR model](HIVPR_dimer.png)

![Figure of 5 chains generated HIV-PR model](HIV_PR_dimer.png)


![Figure of top validation model generated HIV-PR model](HIV_pr_validation_top_model.png)

![Figure of bottom validation model generated HIV-PR](HIV_pr_dimer_validation_model 5.png)

## Custom analysis of resulting models


Read key result files into R. The first thing I need to know is what my results directory/folder is called (i.e. it's name is different for every AlphaFold run/job). 

```{r}
results_dir <- "HIVPR_dimer_23119/"

# File names for all PDB models
pdb_files <- list.files(path=results_dir,
                        pattern="*.pdb",
                        full.names = TRUE)

# Print our PDB file names
basename(pdb_files)
```

```{r}
library(bio3d)

m1 <- read.pdb(pdb_files[1])
m1
```

```{r}
head(m1$atom)
```

```{r}
m1$atom$b[m1$calpha]
```

```{r}
plot(m1$atom$b[m1$calpha], typ="l", ylim=c(0,100))
```

```{r}
plot.bio3d(m1$atom$b[m1$calpha], typ="l")
```


## Residue conservation from alignment file

```{r}
aln_file <- list.files(path=results_dir,
                       pattern=".a3m$",
                        full.names = TRUE)
aln_file
```
Read this into R
```{r}
aln <- read.fasta(aln_file[1], to.upper = TRUE)
```
How many sequences are in this alignment

```{r}
dim(aln$ali)
```

We can score residue conservation in the alignment with the conserv() function.

```{r}
# Read a reference PDB structure
pdb <- read.pdb("1hsg")
sim <- conserv(aln)
plotb3(sim[1:99], sse=trim.pdb(pdb, chain="A"),
       ylab="Conservation Score")
```

Note the conserved Active Site residues D25, T26, G27, A28. These positions will stand out if we generate a consensus sequence with a high cutoff value:

```{r}
con <- consensus(aln, cutoff = 0.9)
con$seq
```






